pkg/repository/v1/payloadstore.go

```diff
@@ -0,0 +1,482 @@
+
+	if err != nil {
+		return false, fmt.Errorf("failed to prepare transaction: %w", err)
+	}
+
+	defer rollback()
+
+	pollLimit := 1000
+	leaseId := uuid.NewString()
+
+	walRecords, err := p.queries.PollPayloadWALForRecordsToOffload(ctx, tx, sqlcv1.PollPayloadWALForRecordsToOffloadParams{
+		Leaseid:         sqlchelpers.UUIDFromStr(leaseId),
+		Polllimit:       int32(pollLimit),
+		Partitionnumber: partitionNumber,
+	})
+
+	hasMoreWALRecords := len(walRecords) == pollLimit
+
+	if len(walRecords) == 0 {
+		return false, nil
+	}
+
+	if err != nil {
+		return false, err
+	}
+
+	retrieveOpts := make([]RetrievePayloadOpts, len(walRecords))
+	retrieveOptsToOffloadAt := make(map[RetrievePayloadOpts]pgtype.Timestamptz)
+
+	for i, record := range walRecords {
+		opts := RetrievePayloadOpts{
+			Id:         record.PayloadID,
+			InsertedAt: record.PayloadInsertedAt,
+			Type:       record.PayloadType,
+			TenantId:   record.TenantID,
+		}
+
+		retrieveOpts[i] = opts
+		retrieveOptsToOffloadAt[opts] = record.OffloadAt
+	}
+
+	payloads, err := p.bulkRetrieve(ctx, tx, retrieveOpts...)
+
+	if err != nil {
+		return false, err
+	}
+
+	externalStoreOpts := make([]StorePayloadOpts, 0)
+	retrieveOptsToPayload := make(map[RetrievePayloadOpts][]byte)
+
+	for opts, payload := range payloads {
+		externalStoreOpts = append(externalStoreOpts, StorePayloadOpts{
+			Id:         opts.Id,
+			InsertedAt: opts.InsertedAt,
+			Type:       opts.Type,
+			Payload:    payload,
+			TenantId:   opts.TenantId.String(),
+		})
+		retrieveOptsToPayload[opts] = payload
+	}
+
+	if err := commit(ctx); err != nil {
+		return false, err
+	}
+
+	retrieveOptsToStoredKey, err := p.offloadToExternal(ctx, externalStoreOpts...)
+
+	if err != nil {
+		return false, err
+	}
+
+	offloadAts := make([]pgtype.Timestamptz, 0, len(retrieveOptsToStoredKey))
+	ids := make([]int64, 0, len(retrieveOptsToStoredKey))
+	insertedAts := make([]pgtype.Timestamptz, 0, len(retrieveOptsToStoredKey))
+	types := make([]string, 0, len(retrieveOptsToStoredKey))
+	values := make([][]byte, 0, len(retrieveOptsToStoredKey))
+	tenantIds := make([]pgtype.UUID, 0, len(retrieveOptsToStoredKey))
+
+	for opt := range retrieveOptsToStoredKey {
+		offloadAt, exists := retrieveOptsToOffloadAt[opt]
+
+		if !exists {
+			return false, fmt.Errorf("offload at not found for opts: %+v", opt)
+		}
+
+		offloadAts = append(offloadAts, offloadAt)
+		ids = append(ids, opt.Id)
+		insertedAts = append(insertedAts, opt.InsertedAt)
+		types = append(types, string(opt.Type))
+
+		key, ok := retrieveOptsToStoredKey[opt]
+		if !ok {
+			return false, fmt.Errorf("external location key not found for opts: %+v", opt)
+		}
+
+		payloadContent := PayloadContent{
+			Location:            PayloadLocationExternal,
+			ExternalLocationKey: &key,
+		}
+
+		marshalledContent, err := payloadContent.Marshal()
+		if err != nil {
+			return false, fmt.Errorf("failed to marshal payload content: %w", err)
+		}
+
+		values = append(values, marshalledContent)
```

Comment:
this is a footgun - since we can write any byte array to the db, it's totally valid Go to just pass the payload itself instead of creating a new `PayloadContent` and then marshalling it